File: rec-core/src/main/java/net/kimleo/rec/repository/RecordSet.java
package net.kimleo.rec.repository;

import net.kimleo.rec.accessor.Accessor;
import net.kimleo.rec.accessor.RecordWrapper;
import net.kimleo.rec.collection.LinkedMultiHashMap;
import net.kimleo.rec.collection.MultiMap;
import net.kimleo.rec.record.Record;
import net.kimleo.rec.sepval.parser.ParseConfig;
import net.kimleo.rec.sepval.parser.SimpleParser;

import java.util.*;
import java.util.function.Function;
import java.util.stream.Collectors;
import java.util.stream.Stream;

import static java.util.Collections.singletonList;
import static java.util.stream.Collectors.toList;
import static java.util.stream.Collectors.toSet;

public class RecordSet implements Iterable<Record> {
    private final List<Record> records;
    private final RecConfig config;
    private final Map<String, MultiMap<String, Record>> indices;
    private final Accessor<String> accessor;

    public Stream<Record> getRecords() {
        return records.stream();
    }

    public RecConfig getConfig() {
        return config;
    }

    public Accessor<String> getAccessor() {
        return accessor;
    }

    public RecordSet(Stream<Record> records, RecConfig config) {
        this.records = records.collect(toList());
        this.config = config;
        this.accessor = config.accessor();
        this.indices = buildIndices(this.records, config.accessor());
    }

    private Map<String, MultiMap<String, Record>> buildIndices(List<Record> records, Accessor<String> accessor) {
        HashMap<String, MultiMap<String, Record>> indices = new HashMap<>();

        for (String key : accessor.getFieldMap().keySet()) {
            LinkedMultiHashMap<String, Record> index = new LinkedMultiHashMap<>();
            records.forEach(rec -> {
                index.put1(accessor.of(rec).get(key), rec);
            });
            indices.put(key, index);
        }
        return indices;
    }

    @Override
    public Iterator<Record> iterator() {
        return records.iterator();
    }

    public RecordSet select(List<String> keys, String name) {
        checkKeyExists(keys);

        RecConfig newType = newType(keys, name);
        Stream<Record> recs = this.records.stream().map(rec -> {
            RecordWrapper<String> acc = accessor.of(rec);
            return new Record(keys.stream().map(acc::get).collect(toList()), rec.parent());
        });

        return new RecordSet(recs, newType);
    }

    public RecordSet select(String... keys) {
        return select(Arrays.asList(keys), null);
    }

    public RecordSet where(String key, Function<String, Boolean> fn) {
        checkKeyExists(singletonList(key));
        Stream<Record> filtered = records.stream().filter(rec -> fn.apply(accessor.of(rec).get(key)));

        return new RecordSet(filtered, config);
    }

    public RecordSet where(Function<RecordWrapper<String>, Boolean> fn) {
        Stream<Record> filtered = records.stream().filter(rec -> fn.apply(accessor.of(rec)));

        return new RecordSet(filtered, config);
    }

    public static RecordSet loadData(Stream<String> lines, RecConfig config) {
        SimpleParser parser = new SimpleParser(config.parseConfig());
        Stream<Record> records = lines
                .map(line -> Record.toRecord(parser.parse(line)));
        return new RecordSet(records, config);
    }

    public boolean contains(String key, String value) {
        checkKeyExists(singletonList(key));
        return indices.get(key).containsKey(value);
    }

    public RecordSet get(String key) {
        return select(key);
    }

    public void verify(Function<RecordWrapper<String>, Boolean> assertion) {
        ArrayList<Record> unexpected = new ArrayList<Record>();

        records.forEach(rec -> {
            if (assertion.apply(accessor.of(rec)))
                unexpected.add(rec);
        });

        unexpected.forEach(rec -> {
            System.out.printf("Unexpected record found: %s", rec);
        });
    }

    private RecConfig newType(List<String> keys, String providedName) {
        return new RecConfig() {
            @Override
            public String name() {
                return providedName != null? providedName : String.format("image of %s", config.name());
            }

            @Override
            public ParseConfig parseConfig() {
                return config.parseConfig();
            }

            @Override
            public String key() {
                return config.key();
            }

            @Override
            public String format() {
                return keys.stream().collect(Collectors.joining(String.valueOf(config.parseConfig().delimiter)));
            }

            @Override
            public Accessor<String> accessor() {
                return new Accessor<>(keys.toArray(new String[] {}));
            }

            @Override
            public boolean keepOrigin() {
                return config.keepOrigin();
            }
        };
    }

    private void checkKeyExists(List<String> keys) {
        if (!indices.keySet().containsAll(keys)) {
            throw new RuntimeException(String.format("Unexpected keys %s", keys));
        }
    }

    public boolean isUnique() {
        return records.size() == records.stream().distinct().count();
    }
}

File: rec-core/src/main/kotlin/net/kimleo/rec/repository/RecordSet.kt
package net.kimleo.rec.repository

import net.kimleo.rec.accessor.Accessor
import net.kimleo.rec.accessor.RecordWrapper
import net.kimleo.rec.bind
import net.kimleo.rec.collection.LinkedMultiHashMap
import net.kimleo.rec.collection.MultiMap
import net.kimleo.rec.orElse
import net.kimleo.rec.record.Cell
import net.kimleo.rec.record.Record
import net.kimleo.rec.record.Record.toRecord
import net.kimleo.rec.sepval.parser.SimpleParser

class RecordSet(val records: List<Record>, val config: RecConfig): Iterable<Record> {

    val indices = buildIndex(records, config.accessor())

    private fun buildIndex(records: List<Record>, accessor: Accessor<String>): Map<String, MultiMap<String, Record>> {
        val indices = hashMapOf<String, MultiMap<String, Record>>()
        for (key in accessor.fieldMap.keys) {
            val index = LinkedMultiHashMap<String, Record>()
            for (record in records) {
                index.put1(accessor.of(record)[key], record)
            }
            indices[key] = index
        }
        return indices
    }

    override fun iterator(): Iterator<Record> {
        return records.iterator()
    }

    val accessor = config.accessor()!!

    fun select(keys: List<String>, name: String? = null): RecordSet {
        checkKeyExists(*keys.toTypedArray())
        val newType = newType(keys, name)
        val newRecords = arrayListOf<Record>()
        for (record in records) {
            val fields = arrayListOf<String>()
            val accessor = accessor.of(record)
            for (key in keys) {
                fields.add(accessor[key].orEmpty())
            }
            newRecords.add(Record(fields.map(::Cell), record.parent()))
        }

        return RecordSet(newRecords, newType)
    }

    fun select(vararg keys: String): RecordSet {
        return select(keys.toList())
    }

    fun where(key: String, assertion: String.() -> Boolean): RecordSet {
        checkKeyExists(key)
        val filtered = records.filter { rec ->
            accessor.of(rec)[key].bind(assertion).orElse { false }
        }
        return RecordSet(filtered, config)
    }

    fun where(assertion: (RecordWrapper<String>) -> Boolean): RecordSet {
        val filtered = records.filter { rec ->
            assertion.invoke(accessor.of(rec))
        }

        return RecordSet(filtered, config)
    }

    companion object {
        fun loadData(lines: List<String>, config: RecConfig): RecordSet {
            val parser = SimpleParser(config.parseConfig())
            val records = arrayListOf<Record>()
            for (line in lines) {
                val record = parser.parse(line)
                record.bind { records.add(toRecord(it)) }
            }

            return RecordSet(records, config)
        }
    }

    fun isUnique(): Boolean {
        return records.toSet().size == records.size
    }

    fun contains(key: String, value: String): Boolean {
        return indices[key].bind { it.containsKey(value) }.orElse { false }
    }


    operator fun get(name: String): RecordSet {
        return select(name)
    }

    private fun checkKeyExists(vararg keys: String) {
        assert(indices.keys.containsAll(keys.asList()))
    }

    private fun newType(keys: List<String>, providedName: String? = null): RecConfig {
        return object: RecConfig {
            override fun name() = providedName.orElse { "image of ${config.name()}" }
            override fun parseConfig() = config.parseConfig()
            override fun key() = config.key()
            override fun format() = keys.joinToString(config.parseConfig().delimiter.toString())
            override fun accessor() = Accessor<String>(keys.toTypedArray())
        }
    }

    fun verify(assertion: (RecordWrapper<String>) -> Boolean) {
        val unexpected = arrayListOf<Record>()
        records.forEach {
            if (!assertion.invoke(accessor.of(it)))
                unexpected.add(it)
        }

        unexpected.forEach {
            println("Unexpected record found [ ${it.text} ]")
        }
    }
}